hs open command

drewjenkins opened the pull request:
Added new `hs open` command.  This command will open a shortcut in the user's web browser.  The list so far isn't the definitive list, and can be added to as we see fit.

There are a few different ways to use `hs open`.

- `hs open`/`hs open --list` - display a list of all available shortcuts
- `hs open shortcut-name` - open the shortcut with name `shortcut-name` in your browser
- `hs open sn` - open the shortcut with an alias of `sn` in your browser
- `hs open settings/navigation` - Some links are sublinks within larger sections.  Stylistically, I used `/` to show this. `settings/navigation` means that we are opening the navigation section within settings.
- `hs open 10` - open the shortcut at index 10 in your browser.  I added this because I see a use case of someone doing `hs open` to find the shortcut they want, then just wanting the fastest way to open it.  They can glance at the index and do this as shorthand.  The index is subject to change if we add more commands, so there is a tiny bit of danger if someone were to do something like `hs open 10` in a script, but I think that risk is pretty negligible and I would hope someone would use the full command in their script anyways.


![2020-10-07 08 24 28](https://user-images.githubusercontent.com/5521743/95330678-e8cf6080-0876-11eb-9a4b-1f9c68cd830f.gif)

anthmatic commented:
Nice! My only thought is that when I ran `hs open` I expected to be able to select a value by index (like and inquirer.js `rawlist` prompt). Even though, as far as I know, we don't use that pattern in this CLI, it might be confusing. That's definitely not a deal breaker, though.

gcorne commented on packages/cms-cli/commands/open.js:
```diff
@@ -0,0 +1,56 @@
+    ['$0 open settings/navigation'],
+    ['$0 open sn'],
+    ['$0 open 10'],
+  ]);
```
Ah, this is cool. We should use it more.

gcorne commented on packages/cms-lib/links.js:
```diff
@@ -0,0 +1,138 @@
+const { getEnv } = require('./lib/config');
```
I think we should keep this in `cms-cli` since it is unlikely to be useful outside the CLI.  Moving to CLI has the added benefit that it reduces the dependencies needed in `@hubspot/cms-lib`.

gcorne commented on packages/cms-lib/links.js:
```diff
@@ -0,0 +1,138 @@
+const { logger } = require('./logger');
+const chalk = require('chalk');
+const { table, getBorderCharacters } = require('table');
+var opn = require('opn');
```
This looks to be deprecated in favor of [open](https://www.npmjs.com/package/open), which we're using elsewhere.

drewjenkins replied:
Ah cool, thanks. I'll change it.

gcorne commented:
> hs open 10 - open the shortcut at index 10 in your browser. I added this because I see a use case of someone doing hs open to find the shortcut they want, then just wanting the fastest way to open it. They can glance at the index and do this as shorthand. The index is subject to change if we add more commands, so there is a tiny bit of danger if someone were to do something like hs open 10 in a script, but I think that risk is pretty negligible and I would hope someone would use the full command in their script anyways.

Maybe I am alone in this thought, but I think having too many ways to accomplish a task is confusing. Is there a reason that we need the numeric form?

drewjenkins commented:
> Maybe I am alone in this thought, but I think having too many ways to accomplish a task is confusing. Is there a reason that we need the numeric form?

The use case was someone that does a `hs open --list` and finds the URL, the quickest way to know what to open is just by glancing at the index, however it is probably unnecessary. I'll remove.

drewjenkins commented:
> Nice! My only thought is that when I ran `hs open` I expected to be able to select a value by index (like and inquirer.js `rawlist` prompt). Even though, as far as I know, we don't use that pattern in this CLI, it might be confusing. That's definitely not a deal breaker, though.

Ah cool, good idea I can look into that

drewjenkins commented:
Thanks for the feedback @gcorne and @anthmatic - incorporated both of your feedback.  Took @gcorne  advice and removed the ability to choose by index, and removed it from `hs open --list`, but took @anthmatic advice and added in Inquirer as a `rawlist` if the user just types `hs open`, which allows interactive selection either through up/down arrows or index.  Seems like the best of both worlds

![2020-10-08 10 34 19](https://user-images.githubusercontent.com/5521743/95473497-2f45bd80-0952-11eb-8cf8-b7a9e33ff00c.gif)

gcorne commented on packages/cms-cli/lib/links.js:
```diff
@@ -0,0 +1,130 @@
+      url: `${baseUrl}/menus/${portalId}/edit/`,
+    },
+    SETTINGS_PAGE: {
+      shortcut: 'settings/page',
```
Maybe `settings/website`?
